agent/agentscripts/agentscripts.go

```diff
@@ -0,0 +1,254 @@
+}
+
+// New creates a runner for the provided scripts.
+func New(ctx context.Context, opts Options) *Runner {
+	return &Runner{
+		Options: opts,
+		cron:    cron.New(cron.WithParser(parser)),
+		ctx:     ctx,
+		closed:  make(chan struct{}),
+	}
+}
+
+type Runner struct {
+	Options
+
+	cmdCloseWait sync.WaitGroup
+	closed       chan struct{}
+	closeMutex   sync.Mutex
+	ctx          context.Context
+	cron         *cron.Cron
+	scripts      []codersdk.WorkspaceAgentScript
+}
+
+// Init initializes the runner with the provided scripts.
+// It also schedules any scripts that have a schedule.
+// This function must be called before Execute.
+func (r *Runner) Init(scripts []codersdk.WorkspaceAgentScript) error {
+	r.scripts = scripts
+
+	for _, script := range scripts {
+		if script.CRON == "" {
+			continue
+		}
+		script := script
+		_, err := r.cron.AddFunc(script.CRON, func() {
+			err := r.run(script)
+			if err != nil {
+				r.Logger.Warn(r.ctx, "run agent script on schedule", slog.Error(err))
+			}
+		})
+		if err != nil {
+			return xerrors.Errorf("add schedule: %w", err)
+		}
+	}
+	return nil
+}
+
+// Execute runs a set of scripts according to a filter.
+func (r *Runner) Execute(filter func(script codersdk.WorkspaceAgentScript) bool) error {
+	if filter == nil {
+		// Execute em' all!
+		filter = func(script codersdk.WorkspaceAgentScript) bool {
+			return true
+		}
+	}
+	var eg errgroup.Group
+	for _, script := range r.scripts {
+		if !filter(script) {
+			continue
+		}
+		script := script
+		eg.Go(func() error {
+			err := r.run(script)
+			if err != nil {
+				return xerrors.Errorf("run agent script %q: %w", script.LogPath, err)
+			}
+			return nil
+		})
+	}
+	return eg.Wait()
+}
+
+// run executes the provided script with the timeout.
+// If the timeout is exceeded, the process is sent an interrupt signal.
+// If the process does not exit after a few seconds, it is forcefully killed.
+// This function immediately returns after a timeout, and does not wait for the process to exit.
+func (r *Runner) run(script codersdk.WorkspaceAgentScript) error {
+	logger := r.Logger.With(slog.F("log_source", script.LogPath))
+	ctx := r.ctx
+	logger.Info(ctx, "running agent script", slog.F("script", script.Source))
+
+	logPath := script.LogPath
+	if logPath == "" {
+		logPath = fmt.Sprintf("coder-%s-script.log", script.LogSourceID)
+	}
+	if !filepath.IsAbs(logPath) {
+		logPath = filepath.Join(r.LogDir, logPath)
+	}
+	fileWriter, err := r.Filesystem.OpenFile(logPath, os.O_CREATE|os.O_RDWR, 0o600)
+	if err != nil {
+		return xerrors.Errorf("open %s script log file: %w", logPath, err)
+	}
+	defer func() {
+		err := fileWriter.Close()
+		if err != nil {
+			logger.Warn(ctx, fmt.Sprintf("close %s script log file", logPath), slog.Error(err))
+		}
+	}()
+
+	var cmd *exec.Cmd
+	if script.Timeout > 0 {
+		var cancel context.CancelFunc
+		// Add a buffer to forcefully kill with the context.
+		ctx, cancel = context.WithTimeout(ctx, script.Timeout+(3*time.Second))
```

Comment:
This may require a rethink. We're passing the agent context in `New` and using it here (`r.ctx`). This means that when the agent is interrupted, all scripts will be terminated, and shutdown scripts won't be able to run.

Currently in `(*agent).Close` we use a background context as base, not the agent context. So I don't think it's a good idea tying the scripts context and agent context together. Instead we should rely on `New` and `Close` to perform the appropriate startup/teardown. In `(agent).Close`, the Execute (stop) should run for as long as it needs to, or until the script timeout is hit (which will interrupt + wait X seconds + kill).

PS. I think this could be bumped to like 5 or 10 seconds at least.